src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/MemoryMarshal.cs

```diff
@@ -382,8 +382,63 @@ public static bool TryGetArray<T>(ReadOnlyMemory<T> memory, out ArraySegment<T>
         /// <returns>An <see cref="IEnumerable{T}"/> view of the given <paramref name="memory" /></returns>
         public static IEnumerable<T> ToEnumerable<T>(ReadOnlyMemory<T> memory)
         {
-            for (int i = 0; i < memory.Length; i++)
-                yield return memory.Span[i];
+            // Extract the object, index, and length.  We need to work on a consistent
+            // view of the struct, which could be torn.
+            object? obj = memory.GetObjectStartLength(out int index, out int length);
+
+            // If the memory is empty, just return an empty array as the enumerable.
+            if (length is 0 || obj is null)
+            {
+                return Array.Empty<T>();
+            }
+
+            // If the object is a string, we can optimize. If it isn't a slice, just return the string as the
+            // enumerable. Otherwise, return an iterator dedicated to enumerating the object; while we could
+            // use the general one for any ReadOnlyMemory, that will incur a .Span access for every element.
+            if (typeof(T) == typeof(char) && obj is string str)
+            {
+                return (IEnumerable<T>)(object)(index == 0 && length == str.Length ?
+                    str :
+                    FromString(str, index, length));
+
+                static IEnumerable<char> FromString(string s, int offset, int count)
+                {
+                    for (int i = 0; i < count; i++)
+                    {
+                        yield return s[offset + i];
+                    }
+                }
+            }
+
+            // If the object is an array, we can optimize. If it isn't a slice, just return the array as the
+            // enumerable. Otherwise, return an iterator dedicated to enumerating the object.
+            if (RuntimeHelpers.ObjectHasComponentSize(obj)) // Same check as in TryGetArray to confirm that obj is a T[] or a U[] which is blittable to a T[].
+            {
+                T[] array = Unsafe.As<T[]>(obj);
+                index &= ReadOnlyMemory<T>.RemoveFlagsBitMask; // the array may be prepinned, so remove the high bit from the start index in the line below.
+                return index == 0 && length == array.Length ?
+                    array :
+                    FromArray(array, index, length);
+
+                static IEnumerable<T> FromArray(T[] array, int offset, int count)
+                {
+                    for (int i = 0; i < count; i++)
+                    {
+                        yield return array[offset + i];
+                    }
+                }
```

Comment:
Maybe return an `ArraySegment<T>` here instead?

Reply:
That would result in a box for the array segment and then a separate enumerator allocation when enumerating it.

Reply:
It seems from this crude benchmark I wrote that looping via `ArraySegment<T>` allocates less, but is also slower (I didn't expect this really). Let me know if I did something wrong, but I think this is worth looking into more.
<details>
<summary>Benchmark code</summary>

```csharp
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<Benchmarks>();

[MemoryDiagnoser(true)]
public class Benchmarks
{
	private int[] intArray = new int[1000];

	[Benchmark]
	public int LoopThrough1()
	{
		int sum = 0;
		foreach (var x in Impl.FromArray(intArray, 0, 1)) sum += x;
		return sum;
	}

	[Benchmark]
	public int LoopThroughViaSegment1()
	{
		int sum = 0;
		foreach (var x in Impl.FromArrayViaSegment(intArray, 0, 1)) sum += x;
		return sum;
	}

	[Benchmark]
	public int LoopThrough1000()
	{
		int sum = 0;
		foreach (var x in Impl.FromArray(intArray, 0, 1000)) sum += x;
		return sum;
	}

	[Benchmark]
	public int LoopThroughViaSegment1000()
	{
		int sum = 0;
		foreach (var x in Impl.FromArrayViaSegment(intArray, 0, 1000)) sum += x;
		return sum;
	}
}

class Impl
{
	public static IEnumerable<T> FromArray<T>(T[] array, int offset, int count)
	{
		for (int i = 0; i < count; i++)
		{
			yield return array[offset + i];
		}
	}
	public static IEnumerable<T> FromArrayViaSegment<T>(T[] array, int offset, int count) => new ArraySegment<T>(array, offset, count);
}
```

</details>

Results:
|                    Method |        Mean |     Error |    StdDev |   Gen0 | Allocated |
|-------------------------- |------------:|----------:|----------:|-------:|----------:|
|              LoopThrough1 |    21.53 ns |  0.228 ns |  0.191 ns | 0.0076 |      64 B |
|    LoopThroughViaSegment1 |    18.84 ns |  0.269 ns |  0.252 ns | 0.0048 |      40 B |
|           LoopThrough1000 | 4,780.79 ns | 77.048 ns | 68.301 ns | 0.0076 |      64 B |
| LoopThroughViaSegment1000 | 5,697.01 ns | 66.120 ns | 55.213 ns |      - |      40 B |

And yes, I just realised I put the number in the wrong spot on the `ViaSegment` ones :)

Reply:
You're not seeing the `IEnumerable<T>` box allocation in your ArraySegment benchmark (I'm guessing because `FromArrayViaSegment` is being inlined and the box eliminated).
If you slap a `[MethodImpl(MethodImplOptions.NoInlining)]` on it your 40B will turn into 72B.

Reply:
Yes, you're right, I've added it to both for this one:
|                    Method |        Mean |     Error |    StdDev |   Gen0 | Allocated |
|-------------------------- |------------:|----------:|----------:|-------:|----------:|
|              LoopThrough1 |    22.38 ns |  0.229 ns |  0.215 ns | 0.0076 |      64 B |
|    LoopThroughViaSegment1 |    27.48 ns |  0.180 ns |  0.159 ns | 0.0086 |      72 B |
|           LoopThrough1000 | 4,566.34 ns | 53.397 ns | 49.947 ns | 0.0076 |      64 B |
| LoopThroughViaSegment1000 | 5,803.10 ns | 48.272 ns | 42.792 ns | 0.0076 |      72 B |